crates/kubelet/src/volume/persistentvolumeclaim.rs

```diff
@@ -0,0 +1,323 @@
+    tokio::fs::write(&vol_data_filepath, serde_json::to_string(&data)?).await?;
+
+    let staging_path = TempDir::new(&csi.volume_handle)?;
+    if stage_unstage_volume {
+        stage_volume(&mut csi_client, &csi, staging_path.path()).await?;
+    }
+    publish_volume(
+        &mut csi_client,
+        &csi,
+        staging_path.path(),
+        stage_unstage_volume,
+        path,
+    )
+    .await?;
+
+    Ok(VolumeType::PersistentVolumeClaim)
+}
+
+// checks if the plugin supports the node_stage/unstage_volume API. Assume false if not specified.
+async fn supports_stage_unstage(
+    csi_client: &mut NodeClient<tonic::transport::Channel>,
+) -> anyhow::Result<bool> {
+    let mut stage_unstage_volume = false;
+    let response = csi_client
+        .node_get_capabilities(NodeGetCapabilitiesRequest {})
+        .await?;
+    for capability in &response.get_ref().capabilities {
+        if let Some(typ) = &capability.r#type {
+            let _typ_stage_unstage_volume = rpc::Type::StageUnstageVolume as i32;
+            match typ {
+                CapabilityType::Rpc(Rpc {
+                    r#type: _typ_stage_unstage_volume,
+                }) => {
+                    stage_unstage_volume = true;
+                }
+            }
+        }
+    }
+    Ok(stage_unstage_volume)
+}
+
+async fn stage_volume(
+    csi_client: &mut NodeClient<tonic::transport::Channel>,
+    csi: &CSIPersistentVolumeSource,
+    staging_path: &Path,
+) -> anyhow::Result<()> {
+    // TODO: grab the publish_context and volume_context using the volume attachments API
+    csi_client
+        .node_stage_volume(NodeStageVolumeRequest {
+            volume_id: csi.volume_handle.clone(),
+            staging_target_path: staging_path.to_string_lossy().to_string(),
+            volume_capability: Some(VolumeCapability {
+                // TODO: determine the correct access mode and mount flags from the volume
+                // https://github.com/kubernetes/kubernetes/blob/734889ed822d1a60c6dd61ccd8f1ed0e8ab31ea5/pkg/volume/csi/csi_attacher.go#L325-L333
+                access_mode: Some(CSIAccessMode {
+                    mode: CSIMode::SingleNodeWriter as i32,
+                }),
+                access_type: Some(CSIAccessType::Mount(CSIMountVolume {
+                    fs_type: csi.fs_type.as_ref().map_or("".to_owned(), |s| s.clone()),
+                    mount_flags: Default::default(),
+                })),
+            }),
+            secrets: Default::default(),
+            publish_context: Default::default(),
+            volume_context: Default::default(),
+        })
+        .await?;
+    Ok(())
+}
+
+async fn publish_volume(
+    csi_client: &mut NodeClient<tonic::transport::Channel>,
+    csi: &CSIPersistentVolumeSource,
+    staging_path: &Path,
+    stage_unstage_volume: bool,
+    path: &PathBuf,
+) -> anyhow::Result<()> {
+    let mut req = NodePublishVolumeRequest {
```

Comment:
Are we supposed to be handling the `NodeUnpublishVolume` request on pod termination? I don't think we are doing that in the providers and so we are leaving volumes published to a node when a pod is no longer there, which will cause errors when they get reused

Reply:
Yeah... I think we need that call. Where would be the best place to implement that call?

Reply:
Looking into implementing this in the `Terminated` state, just before calling `state_reader.stop()`.

Reply:
@bacongobbler I was actually thinking it might be better to hook something in when the state machine terminates or to introduce a new common state called `volume cleanup`

Reply:
IMO we should try not to add this to the state machine. It would be preferable that providers which do not use the common states do not have to know / remember to re-implement this.

Reply:
I hooked this into the `Terminated` state for the time being. Love you both but I cannot tell which direction y'all want me to take 😂 

Can you two please converse and let me know what your decision is after working through the design?

The nice thing is that I was able to refactor a bit of the volume package. Most of it should easily hook into whichever system y'all want me to write.

Reply:
After talking this over offline, we are going to stick with having the cleanup be in the `Terminated` state for now. However, we realized that Krator needs some sort of a deregistration hook and that has been captured in #504

Reply:
Okay! I added a note in the code along with a link to the ticket. Thanks everyone.